File: practica 1/sesion 2/DS_P1S2-FactoriaAbstracta-MetodoFactoriaJava/src/ds_p1_sesion2/Trabajo.java
package ds_p1_sesion2;
import java.util.Vector;

public class Trabajo {
    private Cliente solicitante;
    private String descripcion;
    private Vector<Tecnico> tecnicosInteresados = new Vector<Tecnico>();
    
    private  float precio;
    private Boolean subastando; 
    
    public Trabajo(Cliente solicitante, String nombre, float precio) {
        this.solicitante = solicitante;
        this.descripcion = nombre;
        this.precio = precio;
        subastando = false;
    }
    
    public Trabajo(FactoriaUsuarios factoria, String nombre_cliente, String nombreTrabajo,  float precio) {
        this.solicitante = factoria.crearCliente(nombre_cliente);
        this.descripcion = nombreTrabajo;
        this.precio = precio;
    }

    @Override
    public String toString() {
        return "Trabajo ---> {" + "Descripción: " + descripcion + ", Presupuesto: " + precio + "}";
    }
    
    public Cliente getSolicitante() {
        return solicitante;
    }

    public void setSolicitante(Cliente solicitante) {
        this.solicitante = solicitante;
    }

    public String getNombre() {
        return descripcion;
    }

    public void setNombre(String nombre) {
        this.descripcion = nombre;
    }

    public Vector<Tecnico> getTecnicosInteresados() {
        return tecnicosInteresados;
    }

    public void setTecnicosInteresados(Vector<Tecnico> _solicita) {
        this.tecnicosInteresados = _solicita;
    }

    public float getPrecio() {
        return precio;
    }

    public void setPrecio(float precio) {
        this.precio = precio;
    }
        
    public void subastar() {
        subastando = true;
        System.out.println("En subasta");
    }

    public void asignar(Tecnico asignado) {
	this.tecnicosInteresados.add(asignado);
        asignado.setTrabajo(this);
    }
    
    // Asignar un nuevo tecnico con una nueva oferta solo si se está en el 
    // periodo de subasta y si la oferta mejora a la anterior, si no mejora
    // se tiene en cuenta al técnico añadiendolo al final de la lista.
    public void recibirPuja(Tecnico nuevoTecnico, float oferta) {
        if (subastando){
            if (oferta < precio) {
               tecnicosInteresados.insertElementAt(nuevoTecnico, 0);
               precio = oferta;
           } else {
               tecnicosInteresados.add(nuevoTecnico);
           }
        }
    }
    
    // Para indicar que se terminó el período de subasta
    public void finSubasta(){
        subastando = false;
        System.out.println("Finaliza la subasta");
        Tecnico ganadorPuja = tecnicosInteresados.get(0); // Cogemos el cero porque el vector está ordenado por ofertas
        ganadorPuja.setTrabajo(this);
    }
}

File: practica 1/sesion 2/DS_P1S2-FactoriaAbstracta-PrototipoRuby/Trabajo.rb
require('./Cliente.rb');
require('./Tecnico.rb');

module DSPrototype
  public
  class Trabajo
  
  	public
  	
  	attr_accessor :solicitante, :nombre, :precio, :solicita
  	
  	def subastar()
  		raise "NYI"
  	end
  
  	public
  	def asignar(tecnicoAsignado)
  		solicita.add(tecnicoAsignado)
  	end
  	
  	def abrir_subasta()
  	  @en_subasta = true
  	  puts "Se comienza a subastar el trabajo " + @nombre
  	end
  	
  	def recibir_puja(tecnico_pujador, presupuesto)
  	  if (@en_subasta == true)
  	    if (presupuesto < @precio)
  	      @precio = presupuesto
  	      @solicita.insert(0, tecnico_pujador)
  	      puts "Recibiendo puja del tecnico #{tecnico_pujador.nombre} su oferta es la mejor: #{presupuesto}" 
  	    else 
  	      solicita.push(tecnico_pujador)
  	      puts "La oferta no fue la mejor que se ha recibido, pero se tendra en cuenta"
  	    end  	    
  	  end
  	end
  
  	def cerrar_subasta
  	  if @en_subasta == true
  	    @en_subasta = false
  	    @solicita[0].set_trabajo(self)
  	    
  	    tecnico_ganador = @solicita[0]
  	    
  	    puts "El tecnico #{tecnico_ganador.nombre} ha ganado la puja"
  	  end
  	end
  	
  	def self.createTrabajoCliente(solicitante, nombre, precio)
  	  return Trabajo.new(solicitante,nombre,precio)
  	end
  	
  	def self.createTrabajoFactoria(factoria, nombre_cliente, nombre_trabajo, precio)
  	  solicitante = factoria.crearCliente(nombre_cliente)
  	  trabajo = Trabajo.new(solicitante, nombre_trabajo, precio)
  	  solicitante._crea.push(trabajo);
  	  return trabajo
  	end
  	
  	def toString()
  	  return "Trabajo > [Solicitante : #{@solicitante.toString()} ; Nombre : #{@nombre} ; Solicitan : #{@solicita} ; Precio : #{@precio}]"
  	end
  	
  	public
  	def initialize(solicitante,nombre,precio)
  		@solicitante = solicitante
  		@nombre = nombre
  		@precio = precio
  		@solicita = Array.new()
  		@en_subasta = false
  	end
  	

  	
  	
  end
end
